vllm/v1/guided_decoding/__init__.py

```diff
@@ -0,0 +1,208 @@
+        self.matcher.reset()
+
+    def __copy__(self):
+        return Grammar(matcher=xgr.GrammarMatcher(
+            self.ctx, max_rollback_tokens=self.max_rollback_tokens),
+                       vocab_size=self.vocab_size,
+                       ctx=self.ctx,
+                       max_rollback_tokens=self.max_rollback_tokens)
+
+
+class GuidedDecodingManager:
+
+    def __init__(self, vllm_config: VllmConfig, max_cache_size: int = 500):
+        tokenizer_group = init_tokenizer_from_configs(
+            model_config=vllm_config.model_config,
+            scheduler_config=vllm_config.scheduler_config,
+            parallel_config=vllm_config.parallel_config,
+            lora_config=vllm_config.lora_config)  # type: ignore[arg-type]
+        tokenizer_group.ping()
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
+        self.vllm_config = vllm_config
+
+        tokenizer = tokenizer_group.get_lora_tokenizer(None)
+        tokenizer_info = xgr.TokenizerInfo.from_huggingface(
+            tokenizer, vocab_size=self.vocab_size)
+        self.compiler = xgr.GrammarCompiler(tokenizer_info, max_threads=8)
+
+        self.max_cache_size = max_cache_size
+        self.request_key_to_grammar: OrderedDict[GuidedDecodingKey,
+                                                 Grammar] = OrderedDict()
+
+        # The default max_workers if not specified is the number of CPUs * 5,
+        # which is way too high since these tasks are CPU-bound, not I/O bound.
+        # We also know we would never dominate CPU usage with just grammar
+        # compilation, so we set it to half the number of CPUs.
+        max_workers = max(1, (multiprocessing.cpu_count() + 1) // 2)
+        self.executor = ThreadPoolExecutor(max_workers=max_workers)
+        self.requests: Set[Request] = set()
+        self._requests_lock = threading.Lock()
+        self.grammar_bitmask = xgr.allocate_token_bitmask(
+            self.vllm_config.scheduler_config.max_num_seqs, self.vocab_size)
+
+    def __getitem__(self, key: GuidedDecodingKey) -> Optional[Grammar]:
+        if key in self.request_key_to_grammar:
+            # Move accessed item to the end (most recently used)
+            value = self.request_key_to_grammar.pop(key)
+            self.request_key_to_grammar[key] = value
+            return value
+        return None
+
+    def remove_requests(self, request_ids: List[str]) -> None:
+        with self._requests_lock:
+            self.requests = {
+                req
+                for req in self.requests if req.request_id not in request_ids
+            }
+
+    def populate_cache(self, request: Request):
+        if not request.use_guided_decoding:
+            return False
+        grammar = self.request_key_to_grammar.get(request.guided_decoding_key)
+        if grammar:
+            request.grammar = copy.copy(grammar)
+            return False
+        request.grammar = self.cache(request)
+        return True
```

Comment:
Do we use the return value? If so, please annotate the return type and add a docstring about it.

Reply:
I removed the return value since it was unused